torch_stoi/stoi.py

```diff
@@ -193,54 +238,93 @@ def detect_silent_frames(x, dyn_range, framelen, hop):
-        # Under 1.8.0
-        except TypeError:
-            return torch.stft(x_padded, fft_size, hop_length=hop, window=win,
-                center=False, win_length=win_len)
+        frames = unfold(x[:, None, None, :], kernel_size=(1, win_len),
+                        stride=(1, hop))
+        # Remove last frame to replicate pystoi behavior
+        # See https://github.com/mpariente/pystoi/issues/32
+        frames = frames[..., :-1]
+        return torch.fft.rfft(frames*win[:, None], n=fft_size, dim=1)
 
     @staticmethod
-    def rowcol_norm(x, mask=None):
+    def rowcol_norm(x):
         """ Mean/variance normalize axis 2 and 1 of input vector"""
         for dim in [2, 1]:
-            x = meanvar_norm(x, mask=mask, dim=dim)
+            x = x - x.mean(dim, keepdim=True)
+            x = x / (x.norm(p=2, dim=dim, keepdim=True) + EPS)
         return x
 
 
-def meanvar_norm(x, mask=None, dim=-1):
-    x = x - masked_mean(x, dim=dim, mask=mask, keepdim=True)
-    x = x / (masked_norm(x, p=2, dim=dim, keepdim=True, mask=mask) + EPS)
-    return x
+def _overlap_and_add(x_frames, hop):
+    batch_size, num_frames, framelen = x_frames.shape
+    # Compute the number of segments, per frame.
+    segments = -(-framelen // hop)  # Divide and round up.
 
+    # Pad the framelen dimension to segments * hop and add n=segments frames
+    signal = pad(
+        x_frames, (0, segments * hop - framelen, 0, segments)
+    )
+
+    # Reshape to a 4D tensor, splitting the framelen dimension in two
+    signal = signal.reshape((batch_size, num_frames + segments, segments, hop))
+    # Transpose dimensions so shape = (batch, segments, frame+segments, hop)
+    signal = signal.permute(0, 2, 1, 3)
+    # Reshape so that signal.shape = (batch, segments * (frame+segments), hop)
+    signal = signal.reshape((batch_size, -1, hop))
 
-def masked_mean(x, dim=-1, mask=None, keepdim=False):
-    if mask is None:
-        return x.mean(dim=dim, keepdim=keepdim)
-    return (x * mask).sum(dim=dim, keepdim=keepdim) / (
-        mask.sum(dim=dim, keepdim=keepdim) + EPS
+    # Now behold the magic!! Remove last n=segments elements from second axis
+    signal = signal[:, :-segments]
+    # Reshape to (batch, segments, frame+segments-1, hop)
+    signal = signal.reshape(
+        (batch_size, segments, num_frames + segments - 1, hop)
     )
+    # This has introduced a shift by one in all rows
+
+    # Now, reduce over the columns and flatten the array to achieve the result
+    signal = signal.sum(axis=1)
+    end = (num_frames - 1) * hop + framelen
+    signal = signal.reshape((batch_size, -1))[:end]
+    return signal
 
 
-def masked_norm(x, p=2, dim=-1, mask=None, keepdim=False):
-    if mask is None:
-        return torch.norm(x, p=p, dim=dim, keepdim=keepdim)
-    return torch.norm(x * mask, p=p, dim=dim, keepdim=keepdim)
+def _mask_audio(x, mask):
+    return torch.stack([
+        pad(xi[mi], (0, 0, 0, len(xi) - mi.sum())) for xi, mi in zip(x, mask)
+    ])
```

Comment:
This is the list comprehension which takes the most time, and it makes sense that it does.